server/src/main/java/org/apache/druid/query/RetryQueryRunner.java

```diff
@@ -0,0 +1,250 @@
+        queryPlus.getQuery().getMostRelevantId()
+    );
+    if (remainingResponses > 0) {
+      throw new ISE("Failed to check missing segments due to missing responds from [%d] servers", remainingResponses);
+    }
+
+    final Object maybeMissingSegments = context.get(ResponseContext.Key.MISSING_SEGMENTS);
+    if (maybeMissingSegments == null) {
+      return Collections.emptyList();
+    }
+
+    return jsonMapper.convertValue(
+        maybeMissingSegments,
+        new TypeReference<List<SegmentDescriptor>>()
+        {
+        }
+    );
+  }
+
+  /**
+   * A lazy iterator populating {@link Sequence} by retrying the query. The first returned sequence is always the base
+   * sequence from the baseQueryRunner. Subsequent sequences are created dynamically whenever it retries the query. All
+   * the sequences populated by this iterator will be merged (not combined) with the base sequence.
+   *
+   * The design of this iterator depends on how {@link MergeSequence} works; the MergeSequence pops an item from
+   * each underlying sequence and pushes them to a {@link java.util.PriorityQueue}. Whenever it pops from the queue,
+   * it pushes a new item from the sequence where the returned item was originally from. Since the first returned
+   * sequence from this iterator is always the base sequence, the MergeSequence will call {@link Sequence#toYielder}
+   * on the base sequence first which in turn initializing query distribution tree. Once this tree is built, the query
+   * servers (historicals and realtime tasks) will lock all segments to read and report missing segments to the broker.
+   * If there are missing segments reported, this iterator will rewrite the query with those reported segments and
+   * reissue the rewritten query.
+   *
+   * @see org.apache.druid.client.CachingClusteredClient
+   * @see org.apache.druid.client.DirectDruidClient
+   */
+  private class RetryingSequenceIterator implements Iterator<Sequence<T>>
+  {
+    private final QueryPlus<T> queryPlus;
+    private final ResponseContext context;
+    private final QueryRunner<T> baseQueryRunner;
+    private final Runnable runnableAfterFirstAttempt;
+
+    private boolean first = true;
+    private Sequence<T> sequence = null;
+    private int retryCount = 0;
+
+    private RetryingSequenceIterator(
+        QueryPlus<T> queryPlus,
+        ResponseContext context,
+        QueryRunner<T> baseQueryRunner,
+        Runnable runnableAfterFirstAttempt
+    )
+    {
+      this.queryPlus = queryPlus;
+      this.context = context;
+      this.baseQueryRunner = baseQueryRunner;
+      this.runnableAfterFirstAttempt = runnableAfterFirstAttempt;
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      if (first) {
+        sequence = baseQueryRunner.run(queryPlus, context);
+        // runnableAfterFirstAttempt is only for testing, it must be no-op for production code.
+        runnableAfterFirstAttempt.run();
```

Comment:
I'm not sure I agree that extending a production class in test code is cleaner than adding a test-only option to production code. They're equivalent in terms of test authenticity, and I'm biased towards thinking that extending things is bad in general (because it splits interconnected logic over two different files). So the test-only option sounds good to me.

Reply:
I agree with @gianm. I'm more worried about that the tests may be able to test different code than production if we forget to sync when we modify this part later in the future.

Reply:
I think the JVM will probably ignore the runnable in production. I thought this only makes the code ugly, didn't think it makes things complicated. If that's the case, I think we can think about how to make it better. Do you think we can do it as a follow-up?

Reply:
Ah, I don't think it makes it overly complex or anything, just more complex than if it wasn't there, so more ugly like you said. I don't think this necessarily even needs to change, I'm just being a hater 😛